builtin/providers/profitbricks/resource_profitbricks_firewall.go

```diff
@@ -0,0 +1,241 @@
+
+	if fw.StatusCode > 299 {
+		return fmt.Errorf("An error occured while creating a firewall rule: %s", fw.Response)
+	}
+
+	err := waitTillProvisioned(meta, fw.Headers.Get("Location"))
+	if err != nil {
+		return err
+	}
+	d.SetId(fw.Id)
+
+	return resourceProfitBricksFirewallRead(d, meta)
+}
+
+func resourceProfitBricksFirewallRead(d *schema.ResourceData, meta interface{}) error {
+	username, password, _ := getCredentials(meta)
+	profitbricks.SetAuth(username, password)
+
+	fw := profitbricks.GetFirewallRule(d.Get("datacenter_id").(string), d.Get("server_id").(string), d.Get("nic_id").(string), d.Id())
+
+	if fw.StatusCode > 299 {
+		return fmt.Errorf("An error occured while fetching a firewall rule  dcId: %s server_id: %s  nic_id: %s ID: %s %s", d.Get("datacenter_id").(string), d.Get("server_id").(string), d.Get("nic_id").(string), d.Id(), fw.Response)
+	}
+
+	d.Set("protocol", fw.Properties.Protocol)
+	d.Set("name", fw.Properties.Name)
+	d.Set("source_mac", fw.Properties.SourceMac)
+	d.Set("source_ip", fw.Properties.SourceIp)
+	d.Set("target_ip", fw.Properties.TargetIp)
+	d.Set("port_range_start", fw.Properties.PortRangeStart)
+	d.Set("port_range_end", fw.Properties.PortRangeEnd)
+	d.Set("icmp_type", fw.Properties.IcmpType)
+	d.Set("icmp_code", fw.Properties.IcmpCode)
+	d.Set("nic_id", d.Get("nic_id").(string))
+
+	return nil
+}
+
+func resourceProfitBricksFirewallUpdate(d *schema.ResourceData, meta interface{}) error {
+	username, password, _ := getCredentials(meta)
+	profitbricks.SetAuth(username, password)
+
+	properties := profitbricks.FirewallruleProperties{}
+
+	if d.HasChange("name") {
+		_, new := d.GetChange("name")
+
+		properties.Name = new.(string)
+	}
+	if d.HasChange("source_mac") {
+		_, new := d.GetChange("source_mac")
+
+		properties.SourceMac = new.(string)
+	}
+	if d.HasChange("source_ip") {
+		_, new := d.GetChange("source_ip")
+
+		properties.SourceIp = new.(string)
+	}
+	if d.HasChange("target_ip") {
+		_, new := d.GetChange("target_ip")
+
+		properties.TargetIp = new.(string)
+	}
+	if d.HasChange("port_range_start") {
+		_, new := d.GetChange("port_range_start")
+
+		properties.PortRangeStart = new.(int)
+	}
+	if d.HasChange("port_range_end") {
+		_, new := d.GetChange("port_range_end")
+
+		properties.PortRangeEnd = new.(int)
+	}
+	if d.HasChange("icmp_type") {
+		_, new := d.GetChange("icmp_type")
+
+		properties.IcmpType = new.(int)
+	}
+	if d.HasChange("icmp_code") {
+		_, new := d.GetChange("icmp_code")
+
+		properties.IcmpCode = new.(int)
+	}
+
+	resp := profitbricks.PatchFirewallRule(d.Get("datacenter_id").(string), d.Get("server_id").(string), d.Get("nic_id").(string), d.Id(), properties)
+
+	if resp.StatusCode > 299 {
+		return fmt.Errorf("An error occured while deleting a firewall rule ID %s %s", d.Id(), resp.Response)
+	}
+
+	err := waitTillProvisioned(meta, resp.Headers.Get("Location"))
+	if err != nil {
+		return err
+	}
+	return nil //resourceProfitBricksFirewallRead(d, meta)
```

Comment:
Update should call Read func as a general rule to make sure all the state values get set as expected